File: frontend/src/services/aiChatService.js
import { apiClient } from './authService';

export const aiChatService = {
    /**
     * Gửi câu hỏi đến AI Chat và nhận phản hồi
     * 
     * @param {Object} data - Chat request data
     * @param {number} data.surveyId - ID của khảo sát (bắt buộc)
     * @param {string} data.questionText - Nội dung câu hỏi (bắt buộc, 1-1000 ký tự)
     * @param {number} [data.userId] - ID của người dùng (tùy chọn)
     * @param {number} [data.topK=5] - Số lượng context documents để retrieve (mặc định: 5)
     * @returns {Promise<Object>} AI Chat response
     */
    processChat: async (data) => {
        try {
            console.log('Calling AI chat:', data);
            const response = await apiClient.post('/ai/chat', data, {
                timeout: 60000 // 60 seconds timeout cho AI chat
            });
            console.log('AI chat response:', response.data);
            return response.data;
        } catch (error) {
            console.error('AI chat error:', error);
            throw error;
        }
    },

    /**
     * Lấy lịch sử chat cho một khảo sát
     * 
     * @param {number} surveyId - ID của khảo sát
     * @param {number} [limit=20] - Số lượng chat logs tối đa (mặc định: 20, tối đa: 100)
     * @returns {Promise<Object>} Chat history response
     */
    getChatHistory: async (surveyId, limit = 20) => {
        try {
            console.log('Getting chat history for survey:', surveyId, 'limit:', limit);
            const response = await apiClient.get(`/ai/chat/history/${surveyId}`, {
                params: { limit }
            });
            console.log('Chat history response:', response.data);
            return response.data;
        } catch (error) {
            console.error('Get chat history error:', error);
            throw error;
        }
    },

    /**
     * Lấy lịch sử chat của một người dùng cụ thể cho một khảo sát
     * 
     * @param {number} surveyId - ID của khảo sát
     * @param {number} userId - ID của người dùng
     * @returns {Promise<Object>} User chat history response
     */
    getUserChatHistory: async (surveyId, userId) => {
        try {
            console.log('Getting user chat history for survey:', surveyId, 'user:', userId);
            const response = await apiClient.get(`/ai/chat/history/${surveyId}/user/${userId}`);
            console.log('User chat history response:', response.data);
            return response.data;
        } catch (error) {
            console.error('Get user chat history error:', error);
            throw error;
        }
    },

    /**
     * Ingest dữ liệu khảo sát vào RAG system
     * 
     * @param {number} surveyId - ID của khảo sát cần ingest
     * @returns {Promise<Object>} Ingestion result
     */
    ingestSurveyData: async (surveyId) => {
        try {
            console.log('Ingesting survey data for RAG:', surveyId);
            const response = await apiClient.post(`/ai/chat/rag/ingest/${surveyId}`, {}, {
                timeout: 120000 // 120 seconds timeout cho ingestion process
            });
            console.log('Survey data ingested:', response.data);
            return response.data;
        } catch (error) {
            console.error('Ingest survey data error:', error);
            throw error;
        }
    },

    /**
     * Kiểm tra sức khỏe của AI Chat service
     * 
     * @returns {Promise<Object>} Health check response
     */
    checkHealth: async () => {
        try {
            const response = await apiClient.get('/ai/chat/health');
            return response.data;
        } catch (error) {
            console.error('AI chat health check error:', error);
            throw error;
        }
    }
};

File: backend/src/main/java/vn/duytan/c1se09/smartsurvey/service/ai/AiChatService.java
package vn.duytan.c1se09.smartsurvey.service.ai;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import vn.duytan.c1se09.smartsurvey.domain.AiChatLog;
import vn.duytan.c1se09.smartsurvey.domain.User;
import vn.duytan.c1se09.smartsurvey.domain.request.ai.AiChatRequestDTO;
import vn.duytan.c1se09.smartsurvey.domain.response.ai.AiChatResponseDTO;
import vn.duytan.c1se09.smartsurvey.repository.AiChatLogRepository;
import vn.duytan.c1se09.smartsurvey.service.UserService;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Service for AI Chat functionality
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AiChatService {

    private final RestTemplate restTemplate;
    private final AiChatLogRepository aiChatLogRepository;
    private final UserService userService;
    private final ObjectMapper objectMapper;

    @Value("${ai.service.base-url:http://localhost:8000}")
    private String aiServiceBaseUrl;

    @Value("${ai.service.timeout:30000}")
    private int requestTimeout;

    /**
     * Process AI chat request
     */
    public AiChatResponseDTO processChat(AiChatRequestDTO request, String username) throws Exception {
        log.info("Processing AI chat for user: {} on survey: {}", username, request.getSurveyId());

        // Validate user exists
        User currentUser = null;
        if (username != null) {
            try {
                currentUser = userService.findUserByEmail(username);
                if (request.getUserId() == null && currentUser != null) {
                    request.setUserId(currentUser.getUserId());
                }
            } catch (Exception e) {
                log.warn("Could not find user: {}", username);
            }
        }

        // Call AI service
        Map<String, Object> aiRequest = createAiChatRequest(request);
        AiChatResponseDTO aiResponse = callAiChatService(aiRequest);

        // Save chat log to database
        AiChatLog chatLog = saveChatLog(request, aiResponse, currentUser);
        aiResponse.setChatId(chatLog.getChatId());

        log.info("AI chat processed successfully with chat_id: {}", chatLog.getChatId());
        return aiResponse;
    }

    /**
     * Get chat history for a survey
     */
    public List<AiChatLog> getChatHistory(Long surveyId, int limit) {
        if (limit > 100)
            limit = 100; // Safety limit
        return aiChatLogRepository.findLatestBySurveyId(surveyId, limit);
    }

    /**
     * Get chat history for a user on specific survey
     */
    public List<AiChatLog> getChatHistoryByUser(Long userId, Long surveyId) {
        return aiChatLogRepository.findByUserIdAndSurveyIdOrderByCreatedAtDesc(userId, surveyId);
    }

    /**
     * Ingest survey data for RAG
     */
    public Map<String, Object> ingestSurveyData(Long surveyId) throws Exception {
        log.info("Ingesting survey data for RAG: {}", surveyId);

        String url = aiServiceBaseUrl + "/ai/rag/ingest/" + surveyId;

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(
                    url, HttpMethod.POST, entity, Map.class);

            @SuppressWarnings("unchecked")
            Map<String, Object> result = response.getBody();
            log.info("RAG ingest completed for survey {}: {}", surveyId, result);
            return result;

        } catch (Exception e) {
            log.error("Error ingesting survey data for RAG: {}", e.getMessage(), e);
            throw new Exception("Failed to ingest survey data: " + e.getMessage());
        }
    }

    private Map<String, Object> createAiChatRequest(AiChatRequestDTO request) {
        Map<String, Object> aiRequest = new HashMap<>();
        aiRequest.put("survey_id", request.getSurveyId());
        aiRequest.put("question_text", request.getQuestionText());
        aiRequest.put("user_id", request.getUserId());
        aiRequest.put("top_k", request.getTopK() != null ? request.getTopK() : 5);
        return aiRequest;
    }

    private AiChatResponseDTO callAiChatService(Map<String, Object> request) throws Exception {
        String url = aiServiceBaseUrl + "/ai/chat";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(request, headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(
                    url, HttpMethod.POST, entity, Map.class);

            @SuppressWarnings("unchecked")
            Map<String, Object> responseBody = response.getBody();

            if (responseBody == null) {
                throw new Exception("Empty response from AI service");
            }

            // Parse response
            Long surveyId = Long.valueOf(responseBody.get("survey_id").toString());
            String questionText = (String) responseBody.get("question_text");
            String answerText = (String) responseBody.get("answer_text");
            Integer topK = Integer.valueOf(responseBody.get("top_k").toString());

            @SuppressWarnings("unchecked")
            List<String> context = (List<String>) responseBody.get("context");

            return AiChatResponseDTO.builder()
                    .surveyId(surveyId)
                    .questionText(questionText)
                    .answerText(answerText)
                    .context(context)
                    .topK(topK)
                    .createdAt(LocalDateTime.now())
                    .build();

        } catch (Exception e) {
            log.error("Error calling AI chat service: {}", e.getMessage(), e);
            throw new Exception("AI service unavailable: " + e.getMessage());
        }
    }

    private AiChatLog saveChatLog(AiChatRequestDTO request, AiChatResponseDTO response, User user) {
        try {
            String contextJson = null;
            if (response.getContext() != null && !response.getContext().isEmpty()) {
                contextJson = objectMapper.writeValueAsString(response.getContext());
            }

            AiChatLog chatLog = AiChatLog.builder()
                    .surveyId(request.getSurveyId())
                    .userId(user != null ? user.getUserId() : request.getUserId())
                    .questionText(request.getQuestionText())
                    .aiResponse(response.getAnswerText())
                    .context(contextJson)
                    .build();

            return aiChatLogRepository.save(chatLog);
        } catch (Exception e) {
            log.error("Error saving chat log: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to save chat log: " + e.getMessage());
        }
    }
}